File: src/lib/wowhead-client/zam-url.ts
import { LRUCache } from 'lru-cache';

import { customFetch } from './http-client';

export type ZamExpansion = 'classic' | 'tbc' | 'wrath' | 'cata' | 'mists' | 'live' | 'ptr' | 'ptr2' | 'latest-available';
export type ZamType = 'npc' | 'object' | 'item' | 'dressing-room' | 'character-customization' | 'itemvisual';

export type BaseZamUrl = {
  expansion: ZamExpansion;
}

export type NpcZamUrl = BaseZamUrl & { type: 'npc', displayId: number };
export type ObjectZamUrl = BaseZamUrl & { type: 'object', displayId: number };
export type DressingRoomZamUrl = BaseZamUrl & { type: 'dressing-room', hash: string };
export type ItemZamUrl = BaseZamUrl & { type: 'item', displayId: number, slotId: number | null };
export type CharacterCustomizationZamUrl = BaseZamUrl & { type: 'character-customization', chrModelId: number };
export type ItemVisualZamUrl = BaseZamUrl & { type: 'itemvisual', visualId: number };
export type ZamUrl = NpcZamUrl | ObjectZamUrl | ItemZamUrl | DressingRoomZamUrl | CharacterCustomizationZamUrl | ItemVisualZamUrl;

export function getZamBaseUrl(expansion: ZamExpansion): string {
  return `https://wow.zamimg.com/modelviewer/${expansion}`;
}

const expansionMap: Record<string, ZamExpansion> = {
  // wowhead expansion -> zam expansion
  classic: 'classic',
  tbc: 'tbc',
  wotlk: 'wrath',
  cata: 'cata',
  'mop-classic': 'mists',
  retail: 'live',
  ptr: 'ptr',
  'ptr-2': 'ptr2',
  '': 'live',
} as const;
const expansions = [...Object.entries(expansionMap)];
const expansionsReverse = [...expansions].reverse();

export async function getZamUrlFromWowheadUrl(url: string): Promise<ZamUrl> {
  const type = getTypeFromWowheadUrl(url);
  if (!type) throw new Error(`Cannot infer type from wowhead url: ${url}`);
  if (type === 'character-customization') {
    throw new Error('Cannot get character customization from wowhead url');
  }
  if (type === 'itemvisual') {
    throw new Error('Cannot get item visual from wowhead url');
  }

  const expansion = getExpansionFromUrl(url) || 'live';
  if (type === 'dressing-room') return { expansion, type, hash: url.split('#')[1].split('?')[0] };

  // npc or item
  const { displayId, slotId } = await getDisplayIdFromUrl(url);
  if (type === 'item') {
    return {
      expansion, type, displayId, slotId,
    };
  }
  return { expansion, type, displayId };
}

export function getExpansionFromUrl(url: string): ZamExpansion | undefined {
  for (const [wowheadEx, zamEx] of expansions) {
    const wowheadPrefix = `https://www.wowhead.com/${wowheadEx ? `${wowheadEx}/` : ''}`;
    const wowZamPrefix = `https://wow.zamimg.com/modelviewer/${zamEx}`;
    if (url.startsWith(wowheadPrefix) || url.startsWith(wowZamPrefix)) return zamEx;
  }
  return undefined;
}

export function getWowheadPrefix(expansion: ZamExpansion): string {
  const [wowheadEx] = expansionsReverse.find(([_, zamEx]) => zamEx === expansion) || [];
  return `https://www.wowhead.com${wowheadEx ? `/${wowheadEx}` : ''}`;
}

export async function getLatestExpansionHavingUrl(path: string): Promise<ZamExpansion> {
  for (const [_, zamEx] of expansionsReverse) {
    try {
      const base = getZamBaseUrl(zamEx);
      const res = await customFetch(`${base}/${path}`);
      if (!res.ok) throw new Error(`Failed to fetch latest available npc: ${res.status} ${res.statusText}`);
      return zamEx;
    } catch (e) {
      // continue
    }
  }
  throw new Error(`Invalid Wowhead URL ${path}`);
}

function getTypeFromWowheadUrl(url: string): ZamType | undefined {
  if (/\/npc[=/]/i.test(url)) return 'npc';
  if (/\/spell[=/]/i.test(url)) return 'npc';
  if (/\/object[=/]/i.test(url)) return 'object';
  if (/\/item[=/]/i.test(url)) return 'item';
  if (/\/dressing-room(\?.+)?[#]/i.test(url)) return 'dressing-room';
  return undefined;
}

function getEntityIdFromWowheadUrl(url: string): number | null {
  const m = url.match(/\/(?:item|npc|object)[=/](\d+)/i);
  return m ? parseInt(m[1], 10) : null;
}

function parseDisplayIdFromGathererHtml(html: string, entityId: number): number | null {
  const match = html.match(/WH\.Gatherer\.addData\([^,]+,[^,]+,\s*(\{[\s\S]*\})\);?/);
  if (!match?.[1]) return null;
  try {
    const data = JSON.parse(match[1]) as Record<string, {
      json?: { displayid?: number };
      jsonequip?: { displayid?: number };
    }>;
    const entry = data[String(entityId)];
    const displayid = entry?.json?.displayid ?? entry?.jsonequip?.displayid;
    return typeof displayid === 'number' ? displayid : null;
  } catch {
    return null;
  }
}

async function fetchItemDisplayIdViaGatherer(expansion: ZamExpansion, itemId: number): Promise<number | null> {
  const gathererUrl = `${getWowheadPrefix(expansion)}/gatherer?items=${itemId}`;
  const res = await customFetch(gathererUrl);
  if (!res.ok) return null;
  return parseDisplayIdFromGathererHtml(await res.text(), itemId);
}

async function getDisplayIdFromUrl(url: string): Promise<{ displayId: number, slotId: number | null }> {
  const byJson = url.match(/\/(?:npc|item)\/(\d+)\.json/i)?.[1];
  if (byJson) return { displayId: parseInt(byJson, 10), slotId: null };

  const html = await fetchWithCache(url);
  const m = html.match(/data-mv-display-id="(\d+)"/);
  if (m) {
    const slotM = html.match(/data-mv-slot="(\d+)"/);
    return { displayId: parseInt(m[1], 10), slotId: slotM ? parseInt(slotM[1], 10) : null };
  }

  const type = getTypeFromWowheadUrl(url);
  const entityId = getEntityIdFromWowheadUrl(url);
  if (type === 'item' && entityId !== null) {
    const fromPage = parseDisplayIdFromGathererHtml(html, entityId);
    if (fromPage !== null) {
      return { displayId: fromPage, slotId: null };
    }
    const expansion = getExpansionFromUrl(url) || 'live';
    const fromGatherer = await fetchItemDisplayIdViaGatherer(expansion, entityId);
    if (fromGatherer !== null) {
      return { displayId: fromGatherer, slotId: null };
    }
  }

  throw new Error(`Cannot extract displayId from wowhead url: ${url}`);
}

const respCache = new LRUCache<string, string>({ max: 200 });
export async function fetchWithCache(url: string): Promise<string> {
  const cached = respCache.get(url);
  if (cached) return cached;
  const res = await customFetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch wowhead url: ${url} ${res.status} ${res.statusText} ${await res.text()}`);
  }
  const text = await res.text();
  respCache.set(url, text);
  return text;
}

File: golang/internal/wowhead/zam_url.go
package wowhead

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// Expansion identifies a zam modelviewer expansion.
type Expansion string

const (
	ExpansionClassic          Expansion = "classic"
	ExpansionTBC              Expansion = "tbc"
	ExpansionWrath            Expansion = "wrath"
	ExpansionCata             Expansion = "cata"
	ExpansionMists            Expansion = "mists"
	ExpansionLive             Expansion = "live"
	ExpansionPTR              Expansion = "ptr"
	ExpansionPTR2             Expansion = "ptr2"
	ExpansionLatestAvailable  Expansion = "latest-available"
)

// ZamType identifies zam URL entity types.
type ZamType string

const (
	ZamTypeNPC          ZamType = "npc"
	ZamTypeObject       ZamType = "object"
	ZamTypeItem         ZamType = "item"
	ZamTypeDressingRoom ZamType = "dressing-room"
)

// NpcURL is a zam NPC meta URL descriptor.
type NpcURL struct {
	Expansion Expansion
	Type      ZamType
	DisplayID int
}

// ObjectURL is a zam object meta URL descriptor.
type ObjectURL struct {
	Expansion Expansion
	Type      ZamType
	DisplayID int
}

// ItemURL is a zam item URL descriptor.
type ItemURL struct {
	Expansion Expansion
	Type      ZamType
	DisplayID int
	SlotID    *int
}

// ZamURL is a parsed zam/wowhead URL.
type ZamURL struct {
	Expansion Expansion
	Type      ZamType
	DisplayID int
	SlotID    *int
	Hash      string
}

var expansionMap = []struct {
	wowhead string
	zam     Expansion
}{
	{"classic", ExpansionClassic},
	{"tbc", ExpansionTBC},
	{"wotlk", ExpansionWrath},
	{"cata", ExpansionCata},
	{"mop-classic", ExpansionMists},
	{"retail", ExpansionLive},
	{"ptr", ExpansionPTR},
	{"ptr-2", ExpansionPTR2},
	{"", ExpansionLive},
}

// GetZamBaseURL returns the zam modelviewer base URL.
func GetZamBaseURL(expansion Expansion) string {
	return "https://wow.zamimg.com/modelviewer/" + string(expansion)
}

// GetWowheadPrefix returns wowhead prefix for an expansion.
func GetWowheadPrefix(expansion Expansion) string {
	for i := len(expansionMap) - 1; i >= 0; i-- {
		if expansionMap[i].zam == expansion {
			if expansionMap[i].wowhead == "" {
				return "https://www.wowhead.com"
			}
			return "https://www.wowhead.com/" + expansionMap[i].wowhead
		}
	}
	return "https://www.wowhead.com"
}

// GetExpansionFromURL parses expansion from a wowhead/zam URL.
func GetExpansionFromURL(url string) Expansion {
	for _, e := range expansionMap {
		prefix := "https://www.wowhead.com/"
		if e.wowhead != "" {
			prefix += e.wowhead + "/"
		}
		zamPrefix := GetZamBaseURL(e.zam)
		if strings.HasPrefix(url, prefix) || strings.HasPrefix(url, zamPrefix) {
			return e.zam
		}
	}
	return ""
}

// GetLatestExpansionHavingURL finds the newest expansion with a meta URL.
func GetLatestExpansionHavingURL(client *HTTPClient, path string) (Expansion, error) {
	for i := len(expansionMap) - 1; i >= 0; i-- {
		zamEx := expansionMap[i].zam
		url := GetZamBaseURL(zamEx) + "/" + path
		res, err := client.Get(url)
		if err != nil {
			continue
		}
		res.Body.Close()
		if res.StatusCode >= 200 && res.StatusCode < 300 {
			return zamEx, nil
		}
	}
	return "", fmt.Errorf("no expansion has url %s", path)
}

// GetZamURLFromWowheadURL parses a wowhead URL into a ZamURL.
func GetZamURLFromWowheadURL(client *HTTPClient, url string) (ZamURL, error) {
	typ := getTypeFromWowheadURL(url)
	if typ == "" {
		return ZamURL{}, fmt.Errorf("cannot infer type from wowhead url: %s", url)
	}
	expansion := GetExpansionFromURL(url)
	if expansion == "" {
		expansion = ExpansionLive
	}
	if typ == ZamTypeDressingRoom {
		hash := strings.Split(strings.Split(url, "#")[1], "?")[0]
		return ZamURL{Expansion: expansion, Type: typ, Hash: hash}, nil
	}
	displayID, slotID, err := getDisplayIDFromURL(client, url)
	if err != nil {
		return ZamURL{}, err
	}
	z := ZamURL{Expansion: expansion, Type: typ, DisplayID: displayID, SlotID: slotID}
	return z, nil
}

func getTypeFromWowheadURL(url string) ZamType {
	switch {
	case regexp.MustCompile(`(?i)/npc[=/]`).MatchString(url), regexp.MustCompile(`(?i)/spell[=/]`).MatchString(url):
		return ZamTypeNPC
	case regexp.MustCompile(`(?i)/object[=/]`).MatchString(url):
		return ZamTypeObject
	case regexp.MustCompile(`(?i)/item[=/]`).MatchString(url):
		return ZamTypeItem
	case regexp.MustCompile(`(?i)/dressing-room(\?.+)?#`).MatchString(url):
		return ZamTypeDressingRoom
	default:
		return ""
	}
}

func getDisplayIDFromURL(client *HTTPClient, url string) (int, *int, error) {
	if m := regexp.MustCompile(`(?i)/(?:npc|item)/(\d+)\.json`).FindStringSubmatch(url); len(m) == 2 {
		id, _ := strconv.Atoi(m[1])
		return id, nil, nil
	}
	html, err := FetchWithCache(client, url)
	if err != nil {
		return 0, nil, err
	}
	if m := regexp.MustCompile(`data-mv-display-id="(\d+)"`).FindStringSubmatch(html); len(m) == 2 {
		id, _ := strconv.Atoi(m[1])
		var slot *int
		if sm := regexp.MustCompile(`data-mv-slot="(\d+)"`).FindStringSubmatch(html); len(sm) == 2 {
			s, _ := strconv.Atoi(sm[1])
			slot = &s
		}
		return id, slot, nil
	}
	if m := regexp.MustCompile(`(?i)/(?:item|npc|object)[=/](\d+)`).FindStringSubmatch(url); len(m) == 2 {
		entityID, _ := strconv.Atoi(m[1])
		if getTypeFromWowheadURL(url) == ZamTypeItem {
			if id, err := ParseDisplayIDFromGathererHTML(html, entityID); err == nil {
				return id, nil, nil
			}
			exp := GetExpansionFromURL(url)
			if exp == "" {
				exp = ExpansionLive
			}
			gathererURL := GetWowheadPrefix(exp) + "/gatherer?items=" + m[1]
			ghtml, err := FetchWithCache(client, gathererURL)
			if err == nil {
				if id, err := ParseDisplayIDFromGathererHTML(ghtml, entityID); err == nil {
					return id, nil, nil
				}
			}
		}
	}
	return 0, nil, fmt.Errorf("cannot extract displayId from wowhead url: %s", url)
}
